EIPS/eip-3712.md

```diff
@@ -0,0 +1,337 @@
+            if(amounts[i] == 0){
+                continue;
+            }
+            allERC20[ids[i]].balances[sender[i]] = allERC20[ids[i]].balances[sender[i]] - amounts[i];
+            allERC20[ids[i]].balances[recipient[i]] = allERC20[ids[i]].balances[recipient[i]] + amounts[i];
+        }
+        emit TransferBatchNoIndexed(ids,sender,recipient,amounts);
+    }
+    
+    function _approve(uint256 id,address owner, address spender, uint256 amount) internal virtual {
+        require(owner != address(0), "ERC20: approve from the zero address");
+        require(spender != address(0), "ERC20: approve to the zero address");
+
+        allERC20[id].allowances[owner][spender] = amount;
+        emit Approval(id,owner, spender, amount);
+    }
+    
+    function _approveBatch(uint256[] memory ids,address owner,address spender,uint256[] memory amounts)internal virtual{
+        require(owner != address(0), "ERC20: approve from the zero address");
+        require(spender != address(0), "ERC20: approve to the zero address");
+        
+        require(ids.length == amounts.length,"ERC20s: length not equal");
+        for(uint256 i = 0;i < ids.length;i++){
+            require(bytes(allERC20[ids[i]].name).length != 0 || ids[i] == 0,"ERC20s: ERC20 not in use");
+            if(amounts[i] == 0){
+                continue;
+            }
+            allERC20[ids[i]].allowances[owner][spender] = allERC20[ids[i]].allowances[owner][spender] - amounts[i];
+            amounts[i] = allERC20[ids[i]].allowances[owner][spender];
+        }
+        
+        emit ApproveBatch(ids,owner,spender,amounts);
+    }
+    
+    function _approveBatch(uint256[] memory ids,address[] memory owner,address spender,uint256[] memory amounts)internal virtual{
+        require(ids.length == amounts.length  && owner.length == amounts.length,"ERC20s: length not equal");
+        
+        for(uint256 i = 0;i < ids.length;i++){
+            require(bytes(allERC20[ids[i]].name).length != 0 || ids[i] == 0,"ERC20s: ERC20 not in use");
+            if(amounts[i] == 0){
+                continue;
+            }
+            allERC20[ids[i]].allowances[owner[i]][spender] = allERC20[ids[i]].allowances[owner[i]][spender] - amounts[i];
+            amounts[i] = allERC20[ids[i]].allowances[owner[i]][spender];
+        }
+        
+        emit ApproveBatchNoIndexed(ids,owner,spender,amounts);
+    }
+    
+    function createMsg(address _msg,uint256 len) internal view returns(address[] memory data){
+        data = new address[](len);
+        for(uint256 i = 0;i < len;i++){
+            data[i] = _msg;
+        }
+    }
+    
+    function _createERC20(uint256 id,string memory name,string memory symbol,uint8 _decimals) internal {
+        require(bytes(allERC20[id].name).length == 0,"ERC20s: ERC20 in use");
+        allERC20[id].name = name;
+        allERC20[id].symbol = symbol;
+        allERC20[id].decimals = _decimals;
+    }
+}
+```
+
```

Comment:
The only blocker that I see left to get this merged is that the `Security Considerations` section is missing and is not optional.